### PR TITLE
when failing to create hard link attempt to use shutil.copy2

### PR DESCRIPTION
**Pull request by matuskosut**
When using e.g. vagrant build fails because using hard links is not allowed.
Could catching the permission exception and attempting to copy the file instead be a way to manage builds in vagrant?
https://github.com/canonical/charmcraft/issues/151

### CODE REVIEW

**Review by facundobatista — CHANGES_REQUESTED** [review]
Hello @matuskosut , thanks for your work here!

The patch looks solid, but we need a test for this case to let it in.

Thanks again!

**Comment by chipaca** [comment]
@matuskosut thanks for this! are you ok if i pick up writing the test for it, or are you already doing so?

**Comment by matuskosut** [comment]
@chipaca sure go ahead, I haven't got back to it yet. Thanks :)

File: tests/commands/test_build.py [hunk 1]
```diff
@@ -596,6 +595,29 @@ def test_build_generics_tree(tmp_path, caplog):
...
+    if expect_hardlinks:
+        # they're hard links
+        assert (build_dir / 'crazycharm.py').samefile(entrypoint)
+        assert (build_dir / 'file1.txt').samefile(file1)
+        assert (build_dir / 'dir2' / 'file2.txt').samefile(file2)
+    else:
+        # they're *not* hard links
+        assert not (build_dir / 'crazycharm.py').samefile(entrypoint)
+        assert not (build_dir / 'file1.txt').samefile(file1)
+        assert not (build_dir / 'dir2' / 'file2.txt').samefile(file2)
```
Comment (facundobatista) [inline tests/commands/test_build.py:607]: Please check that the files has the same content, permissions, etc... thanks!

**Review by facundobatista — CHANGES_REQUESTED** [review]
If we copy the files, we need to verify they are properly copied.

File: tests/commands/test_build.py (hunk 1 above)
Comment (chipaca) [reply tests/commands/test_build.py:607]: picky, picky...! :-D

**Review by facundobatista — APPROVED** [review]
Wonderful, thanks

**Review by balbirthomas — APPROVED** [review]

**Pull request merged**